src/extractor/extractor.cpp

```diff
@@ -808,5 +818,173 @@ void Extractor::WriteCompressedNodeBasedGraph(const std::string &path,
+  {
+    bool operator() (EdgeInfo const & e1, EdgeInfo const & e2) const
+    {
+      return e1.name_id < e2.name_id;
+    }
+  };
+};
+
+bool IsSegregated(std::vector<EdgeInfo> v1, std::vector<EdgeInfo> v2, EdgeInfo const & current)
+{
+  if (v1.size() < 2 || v2.size() < 2)
+    return false;
+
+  auto const preprocessFn = [] (std::vector<EdgeInfo> & v)
+  {
+    std::sort(v.begin(), v.end(), EdgeInfo::LessName());
+  };
+
+  preprocessFn(v1);
+  preprocessFn(v2);
+
+  // Internal edge with the name should be connected with any other neibour edge with the same name.
+  if (current.name_id != EMPTY_NAMEID)
+  {
+    auto const findNameFn = [&current] (std::vector<EdgeInfo> const & v)
+    {
+      return std::binary_search(v.begin(), v.end(), current, EdgeInfo::LessName());
+    };
+
+    if (!findNameFn(v1) && !findNameFn(v2))
+      return false;
+  }
+
+  // set_intersection like routine to count equal name pairs.
+  std::vector<std::pair<EdgeInfo const *, EdgeInfo const *>> commons;
+
+  auto i1 = v1.begin();
+  auto i2 = v2.begin();
+
+  while (i1 != v1.end() && i2 != v2.end())
+  {
+    if (i1->name_id == i2->name_id)
+    {
+      if (i1->name_id != EMPTY_NAMEID)
+        commons.push_back(std::make_pair(&(*i1), &(*i2)));
+
+      ++i1;
+      ++i2;
+    }
+    else if (i1->name_id < i2->name_id)
+      ++i1;
+    else
+      ++i2;
+  }
+
+  return (commons.size() >= 2);
+
+  /// @todo Process standalone U-turns.
+//  switch (commons.size())
+//  {
+//  case 0:
+//    return false;
+//  case 1:
+//    // ingoing + outgoing edges
+//    if (commons.front().first->direction + commons.front().second->direction != 1)
+//      return false;
+//  }
+
+//  return true;
+}
+
+void Extractor::FindSegregatedNodes(std::shared_ptr<util::NodeBasedDynamicGraph> graph,
+                                    std::vector<util::Coordinate> const & coordinates)
+{
+  auto const collectEdgeInfoFn = [&graph](auto const & edges1, NodeID node2)
+  {
+    std::vector<EdgeInfo> info;
+
+    for (auto const & e : edges1)
+    {
+      NodeID const target = graph->GetTarget(e);
+      if (target == node2)
+        continue;
+
+      auto const & data = graph->GetEdgeData(e);
+
+      BOOST_ASSERT(data.name_id != INVALID_NAMEID);
+      info.push_back({target, data.name_id, data.reversed ? 1 : 0, data.classes});
+    }
+
+    if (info.empty())
+      return info;
+
+    std::sort(info.begin(), info.end(), [](EdgeInfo const & e1, EdgeInfo const & e2)
+    {
+      return e1.node < e2.node;
+    });
+
+    // std::unique like code with custom combine routine.
+    auto curr = info.begin();
+    auto next = curr;
+    while (++next != info.end())
+    {
+      if (curr->node != next->node)
+      {
+        *(++curr) = *next;
+      }
+      else
+      {
+        BOOST_ASSERT(curr->name_id == next->name_id);
+        BOOST_ASSERT(curr->road_class == next->road_class);
+
+        BOOST_ASSERT(curr->direction != next->direction);
+        curr->direction = 2;
```

Comment:
could we split this into two runs to find which ones are bidirecitonal (I guess that is was you are shooting for here?) and then use std::unique?